src/notebooks/controllers/vscodeNotebookController.ts

```diff
@@ -434,13 +462,59 @@ export class VSCodeNotebookController implements Disposable {
 
         // Execution should be ended elsewhere
     }
+    private static readonly connections = new WeakMap<
+        NotebookDocument,
+        { kernel: Deferred<IKernel>; options: IDisplayOptions }
+    >();
+    private async connectToKernel(
+        doc: NotebookDocument,
+        options: IDisplayOptions = new DisplayOptions(false)
+    ): Promise<IKernel> {
+        let currentPromise = VSCodeNotebookController.connections.get(doc);
```

Comment:
**This is one of the bugs that was fixed, its an easy to repro issue:**
* Open a notebook, select a kernel A without ipykernel
* Run all cells
* In the prompt displayed select the option to change the kernel to B
* We still get error message indicating ipykernel isnt found in A
* Run all cells again, and we'll still get errors about ipykernel not in A

The problem was we had `connectToKernel` function called from different places and they all had their own while loops, and thats what caused issues (causing multiple kernels to get created, one gets created and disposes the other, and even after we change the controller, the previous loop will treat the older controller as active and weird things happend).

Basically we were running `connectToKernel` function and had multiple while loops (instead of one) = now we cache it per doc.